storage/local/storage.go

```diff
@@ -505,37 +516,51 @@ func (s *memorySeriesStorage) MetricsForLabelMatchers(matchers ...*metric.LabelM
 	return result
 }
 
-// LabelValuesForLabelName implements Storage.
-func (s *memorySeriesStorage) LabelValuesForLabelName(labelName model.LabelName) model.LabelValues {
-	lvs, err := s.persistence.labelValuesForLabelName(labelName)
-	if err != nil {
-		log.Errorf("Error getting label values for label name %q: %v", labelName, err)
-	}
-	return lvs
-}
-
-// MetricForFingerprint implements Storage.
-func (s *memorySeriesStorage) MetricForFingerprint(fp model.Fingerprint) metric.Metric {
-	s.fpLocker.Lock(fp)
-	defer s.fpLocker.Unlock(fp)
-
+// metricForRange returns the metric for the given fingerprint if the
+// corresponding time series has samples between 'from' and 'through', together
+// with a pointer to the series if it is in memory already. For a series that
+// does not have samples between 'from' and 'through', the returned bool is
+// false. For an archived series that does contain samples between 'from' and
+// 'through', it returns (metric, nil, true).
+//
+// The caller must have locked the fp.
+func (s *memorySeriesStorage) metricForRange(
+	fp model.Fingerprint,
+	from, through model.Time,
+) (model.Metric, *memorySeries, bool) {
 	series, ok := s.fpToSeries.get(fp)
 	if ok {
-		// Wrap the returned metric in a copy-on-write (COW) metric here because
-		// the caller might mutate it.
-		return metric.Metric{
-			Metric: series.metric,
+		if series.lastTime.Before(from) || series.firstTime().After(through) {
+			return nil, nil, false
+		}
+		return series.metric, series, true
```

Comment:
The problem is that the function _only_ returns the metric if the series isn't easily accessible (namely if it's archived).  See doc comment and line 558.

This all is the result of the weird flow. For one codepath, we only need the metric, for another we need the series. For the former, we really don't want to access the archive index to get the series we don't need anyway. For the former, we do want to return the series if we already have it anyway to not do the relatively expensive synchronized map lookup again. But otherwise, it's exactly the same logic, which was completely duplicated before.